contracts/StorageManager.sol

```diff
@@ -172,177 +156,180 @@ contract StorageManager {
+        } else {
+            // Consumer can withdraw all funds except for those already used for past storage hosting
+            // AND for current period
+            maxWithdrawableFunds = agreement.availableFunds - _calculateSpentFunds(agreement) - (agreement.billingPrice * agreement.size);
+        }
+
+        if (amount == 0) {
+            amount = maxWithdrawableFunds;
+        }
+
+        require(amount <= maxWithdrawableFunds, "StorageManager: Amount is too big");
+        agreement.availableFunds = agreement.availableFunds.sub(amount);
+        msg.sender.transfer(amount);
+        emit AgreementFundsWithdrawn(agreementReference, amount);
     }
 
     /**
+    >> FOR PROVIDER
     @notice payout already earned funds of one or more Agreement
     @dev 
-    - any safeMath operations in this function don't cause a deadlock, as a possible revert just means we have to add less agreementReferences
-    - Provider must call an expired agreement themselves as soon as the agreement is expired, to add back the capacity.
+    - Provider must call an expired agreement themselves as soon as the agreement is expired, to add back the capacity to their Offer.
     @param agreementReferences reference to one or more Agreement
     */
     function payoutFunds(bytes32[] memory agreementReferences) public {
+        Offer storage offer = offerRegistry[msg.sender];
+        require(offer.availableCapacity != 0, "StorageManager: Offer for this Provider doesn't exist");
         uint256 toTransfer;
-        for(uint8 i = 0; i < agreementReferences.length; i++) {
-            Agreement storage agreement = offerRegistry[msg.sender].agreementRegistry[agreementReferences[i]];
-            require(agreement.startDate != 0, "StorageManager: Agreement not active");
-            // check if agreement is expired
-            //NO_OVERFLOW REASONING: see REF_DURATION
-            if((agreement.startDate + agreement.numberOfPeriodsDeposited) * agreement.billingPeriod < now) {
-                //NO_OVERFLOW reasoning: numberOfPeriodsWithdrawn is always less than or equal to numberOfPeriodsDeposited and REF_MAX_TRANSFER
-                toTransfer = toTransfer.add((agreement.numberOfPeriodsDeposited - agreement.numberOfPeriodsWithdrawn) * agreement.billingPrice);
-                //reset agreement to clean state
-                agreement.numberOfPeriodsWithdrawn = 0;
-                agreement.numberOfPeriodsDeposited = 0;
-                agreement.startDate = 0;
-                // check if Offer is still active
-                if(offerRegistry[msg.sender].availableCapacity != 0) {
+        for (uint8 i = 0; i < agreementReferences.length; i++) {
+            Agreement storage agreement = offer.agreementRegistry[agreementReferences[i]];
+            require(agreement.size != 0, "StorageManager: Agreement for this Offer doesn't exist");
+
+            // TODO: [Q] Should we disallow to payout from inactive agreement? Can it cause some deadlock?
```

Comment:
[Q] Should we automatically return any remaining agreement.availableFunds?

Reply:
I think we could. It would be nice from the provider.

Reply:
The Consumer also maybe will return to the expired Agreement in such a case the previous funds would be only added to the newly deposited ones.

Reply:
This is regarding the "Consumer" funds which are still locked in an expired Agreement? I think we can return them directly.

Reply:
I have realized that in this method I don't have Consumer's address, so it is not really possible in the current structure and I don't feel implementing this is so crucial in order to change the Agreement to support this.